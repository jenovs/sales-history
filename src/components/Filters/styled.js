import styled from '@emotion/styled';

export const Group = styled.div`
  margin-bottom: 1rem;
`;

export const H2 = styled.h2`
  color: #333;
  font-size: 1.3rem;
  margin-bottom: 0.5rem;
`;

export const Overlay = styled.div`
  display: none;
  position: absolute;
  z-index: -1;
  @media (pointer: coarse) {
    align-items: center;
    color: #777;
    display: flex;
    font-family: 'monospace';
    font-size: 4rem;
    font-weight: bold;
    height: 100%;
    justify-content: center;
    width: 100%;
  }
`;

export const Wrapper = styled.div`
  padding-left: 3rem;
  padding-top: 3rem;
  max-width: 500px;
  position: relative;
  width: 100%;

  @media (max-width: 900px) {
    padding-top: 0;
    padding-left: 0;
  }

  @media (max-width: 350px) {
    font-size: 0.8rem;
  }
`;
