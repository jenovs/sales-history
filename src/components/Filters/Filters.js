import React from 'react';
import PropTypes from 'prop-types';

import Button from '../Button';
import { Group, H2, Overlay, Wrapper } from './styled';
import Loading from '../Loading';

const propTypes = {
  filterConfig: PropTypes.array.isRequired,
  filterState: PropTypes.object.isRequired,
  loading: PropTypes.bool,
  updateFilter: PropTypes.func.isRequired,
};

const defaultProps = {
  loading: false,
};

// eslint-disable-next-line
const Filters = ({ filterConfig, filterState, updateFilter, loading }) => {
  const filterGroups = [...new Set(filterConfig.map(f => f.filterGroupLabel))];

  return (
    <Wrapper>
      {loading && (
        <Overlay>
          <Loading />
        </Overlay>
      )}
      {filterGroups.map(group => (
        <Group key={group}>
          <H2>{group}</H2>
          {filterConfig.reduce((acc, g) => {
            if (g.filterGroupLabel === group) {
              acc.push(
                <Button
                  key={g.name}
                  onClick={() => updateFilter(g.name)}
                  inactive={!filterState[g.name]}
                >
                  {g.label}
                </Button>
              );
            }
            return acc;
          }, [])}
        </Group>
      ))}
    </Wrapper>
  );
};

Filters.propTypes = propTypes;
Filters.defaultProps = defaultProps;

export default Filters;
