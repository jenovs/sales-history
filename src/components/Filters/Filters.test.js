import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import renderer from 'react-test-renderer';
import 'jest-dom/extend-expect';

import Filters from '.';

const mockFn = jest.fn();

const props = {
  filterConfig: [
    {
      name: 'card',
      label: 'Card',
      filterGroupLabel: 'Payment type',
      filterGroupName: 'paidBy',
    },
    {
      name: 'cash',
      label: 'Cash',
      filterGroupLabel: 'Payment type',
      filterGroupName: 'paidBy',
    },
    {
      name: 'successful',
      label: 'Successful',
      filterGroupLabel: 'Status',
      filterGroupName: 'status',
    },
  ],
  filterState: { card: false, cash: true, successful: true },
  updateFilter: mockFn,
};

describe('Filter', () => {
  afterEach(cleanup);

  it('should render two headers and three filter buttons', () => {
    const { getByText } = render(<Filters {...props} />);

    expect(getByText('Payment type')).toBeVisible();
    expect(getByText('Card')).toBeVisible();
    expect(getByText('Cash')).toBeVisible();
    expect(getByText('Status')).toBeVisible();
    expect(getByText('Successful')).toBeVisible();
  });

  it('should handle click on buttons', () => {
    const { getByText } = render(<Filters {...props} />);

    fireEvent.click(getByText('Card'));
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledWith('card');

    mockFn.mockClear();

    fireEvent.click(getByText('Cash'));
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledWith('cash');

    mockFn.mockClear();

    fireEvent.click(getByText('Successful'));
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(mockFn).toHaveBeenCalledWith('successful');
  });

  it('should render buttons with different state', () => {
    const tree = renderer.create(<Filters {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
