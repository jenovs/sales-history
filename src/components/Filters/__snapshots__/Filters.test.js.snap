// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should render buttons with different state 1`] = `
<div
  className="css-1bmxqqd"
>
  <div
    className="css-wndajv"
  >
    <h2
      className="css-1xblkvf"
    >
      Payment type
    </h2>
    <button
      className="css-1jabc46"
      onClick={[Function]}
      type="button"
    >
      Card
    </button>
    <button
      className="css-1lnopfn"
      onClick={[Function]}
      type="button"
    >
      Cash
    </button>
  </div>
  <div
    className="css-wndajv"
  >
    <h2
      className="css-1xblkvf"
    >
      Status
    </h2>
    <button
      className="css-1lnopfn"
      onClick={[Function]}
      type="button"
    >
      Successful
    </button>
  </div>
</div>
`;
